subsys/net/l2/ppp/ppp_l2.c

```diff
@@ -195,133 +195,30 @@ static void ppp_close(struct ppp_context *ctx)
 {
 	if (ppp_lcp) {
 		ppp_lcp->close(ctx, "Shutdown");
+	} else {
+		ppp_change_phase(ctx, PPP_DEAD);
 	}
 }
 
-static void ppp_lower_up(struct ppp_context *ctx)
-{
-	if (ppp_lcp) {
-		ppp_lcp->lower_up(ctx);
-	}
-}
-
-static void start_ppp(struct ppp_context *ctx)
+static void ppp_open(struct ppp_context *ctx)
 {
 	ppp_change_phase(ctx, PPP_ESTABLISH);
 
-	ppp_lower_up(ctx);
-
 	if (ppp_lcp) {
 		NET_DBG("Starting LCP");
+		ppp_lcp->lower_up(ctx);
 		ppp_lcp->open(ctx);
 	}
 }
 
-static int ppp_enable(struct net_if *iface, bool state)
-{
-	const struct ppp_api *ppp =
-		net_if_get_device(iface)->api;
-	struct ppp_context *ctx = net_if_l2_data(iface);
-
-
-	if (ctx->is_enabled == state) {
-		return 0;
-	}
-
-	ctx->is_enabled = state;
-
-	if (!state) {
-		ppp_close(ctx);
-
-		if (ppp->stop) {
-			ppp->stop(net_if_get_device(iface));
-		}
-	} else {
-		if (ppp->start) {
-			ppp->start(net_if_get_device(iface));
-		}
-
-		if (ctx->is_startup_pending) {
-			ctx->is_enable_done = true;
-		} else {
-			start_ppp(ctx);
-		}
-	}
-
-	return 0;
-}
-
 static enum net_l2_flags ppp_flags(struct net_if *iface)
 {
 	struct ppp_context *ctx = net_if_l2_data(iface);
 
 	return ctx->ppp_l2_flags;
 }
 
-NET_L2_INIT(PPP_L2, ppp_recv, ppp_send, ppp_enable, ppp_flags);
-
-/* A workaround for PPP L2 not yet supporting net_if_carrier_on/off(). */
-void ppp_if_carrier_down(struct net_if *iface)
-{
-	net_if_flag_clear(iface, NET_IF_UP);
-	net_mgmt_event_notify(NET_EVENT_IF_DOWN, iface);
-	net_ipv4_autoconf_reset(iface);
-}
-
-static void carrier_on_off(struct k_work *work)
-{
-	struct ppp_context *ctx = CONTAINER_OF(work, struct ppp_context,
-					       carrier_work);
-	bool ppp_carrier_up;
-
-	if (ctx->iface == NULL) {
-		return;
-	}
-
-	ppp_carrier_up = atomic_test_bit(&ctx->flags, PPP_CARRIER_UP);
-
-	if (ppp_carrier_up == (bool) ctx->is_net_carrier_up) {
-		return;
-	}
-
-	ctx->is_net_carrier_up = ppp_carrier_up;
-
-	NET_DBG("Carrier %s for interface %p", ppp_carrier_up ? "ON" : "OFF",
-		ctx->iface);
-
-	if (ppp_carrier_up) {
-		ppp_mgmt_raise_carrier_on_event(ctx->iface);
-		net_if_up(ctx->iface);
-	} else {
-		if (ppp_lcp) {
-			ppp_lcp->close(ctx, "Shutdown");
-			/* signaling for the carrier off event is done from the LCP callback */
-		} else {
-			ppp_change_phase(ctx, PPP_DEAD);
-
-			ppp_mgmt_raise_carrier_off_event(ctx->iface);
-			ppp_if_carrier_down(ctx->iface);
-		}
-	}
-}
-
-void net_ppp_carrier_on(struct net_if *iface)
-{
-	struct ppp_context *ctx = net_if_l2_data(iface);
-
-	if (!atomic_test_and_set_bit(&ctx->flags, PPP_CARRIER_UP)) {
-		k_work_submit(&ctx->carrier_work);
-	}
-}
-
-void net_ppp_carrier_off(struct net_if *iface)
-{
-	struct ppp_context *ctx = net_if_l2_data(iface);
-
-	if (atomic_test_and_clear_bit(&ctx->flags, PPP_CARRIER_UP)) {
-		k_work_submit(&ctx->carrier_work);
-	}
-}
+NET_L2_INIT(PPP_L2, ppp_recv, ppp_send, NULL, ppp_flags);
```

Comment:
The PPP L2 implementation can not handle either ADMIN UP/DOWN, CARRIER ON/OFF or DORMANT ON/OFF states at all currently, it only handles IF_UP/DOWN (ESTABLISH/CLOSE session) which are a result of a combination of the states. Therefore there is no work to be done on any event outside of NET_IF_UP/DOWN. It would require extensive changes to the internals of the PPP L2 implementation to add this, which is outside the scope of this PR.

Reply:
As CARRIER/DORMANT ON/OFF are not handled by PPP L2, it's not clear to me, why would we need to regsiter for `NET_IF_UP/DOWN` events, instead of calling `ppp_open/close()` from the `enable()` handler. Maybe I overlook something, so please exaplain in that case, but it seems more conceptually correct, and leaves opening for future imporovements in this area.

Reply:
The `NET_IF_UP` event occurs when: `NET_IF_ADMIN_UP && NET_IF_CARRIER_ON && NET_IF_DORMANT_OFF`
The `NET_IF_DOWN` event occurs when `!NET_IF_ADMIN_UP || !NET_IF_CARRIER_ON || !NET_IF_DORMANT_OFF`

Using these events, a combination of `net_if_up()`, `net_if_carrier_on()` and `net_if_dormant_off()` will bring up the interface as expected, any call to `net_if_down()` from the app, or `net_if_carrier_off()` or `net_if_dormant_on()` from drivers will bring down the network.

This demonstrates the expected behavior from the application and drivers, using the net_if interface.

This solution does leave the door open for future improvements internally in the NET L2 implementation to actually react to the specific underlying event that caused `NET_IF_UP/DOWN`, transparently to the apps/drivers interacting with the net_if.

Reply:
The issue with using `ppp_open/close()` from the `enable()` function is that this occurs synchronously with the call to `net_if_up/down()`, at which point, the net_if could be dormant, or have no carrier, calling `ppp_start()` at this time will not work as ppp_start() expects the interface to be ready to receive packets. It will work for  `ppp_stop()`, but since this is already called when the `NET_IF_DOWN` occurs, it is redundant in this case

Reply:
`net_if_dormant_on()` yes, however, the `net_if_carrier_off()` is caused by a physical event, like a cable being disconnected, or for modems, loosing connection to the cellular network, which is triggered by the drivers (L1). The PPP L2 implementation does not use or implement to the `net_if_dormant_on()` at all currently, so if it is called from outside the PPP L2 implementation (which it should not be), it will just bring down the interface.

This PR only deals with the interactions between app/drivers and the PPP L2 implementation, so `net_if_up/down()` and `net_if_carrier_on/off()`

Reply:
Ok, thanks for all of the explanations. I have an impression we're missing `NET_EVENT_IF_CARRIER_ON/OFF` event that would be useful for this case, so that L2 could detect carrier changes specifically, instead of relying on `NET_IF_UP/DOWN` events.

But that's probably out of scope of this PR, so I'm not pushing this further.